### PR TITLE
Ensure that the index processed by the client is at least as new as the last one processed.

### PR DESCRIPTION
**Pull request by stswidwinski**
A fix for https://github.com/hashicorp/nomad/issues/18267, ensuring that "time travel" of the client state is not possible.

I could not find tests which may exercise this logic meaningfully. Please point me the right direction and I'm happy to add them

Please note that this isn't an actual good fix to the underlying issue, but rather a stop-gap measure.

### CODE REVIEW

**Comment by stswidwinski** [comment]
I don't think the UI tests are related here. May I ask someone to re-run the tests or help me understand the failure mode?

**Comment by lgfa29** [comment]
Thanks for the PR and detailed issue description @stswidwinski.

> I could not find tests which may exercise this logic meaningfully. Please point me the right direction and I'm happy to add them

Yeah, this part of the code is very tricky to test reliably since there's a lot going on, causing tests to be flaky sometimes. [`TestClient_WatchAllocs`](https://github.com/hashicorp/nomad/blob/14a38bee7bc4386e74157f6a99f3db7382d7e6a5/client/client_test.go#L539-L625) may be a good place to start. I wonder if you setup a test like that and manually mess with the server raft index we could reliably reproduce and test the problem.

> I don't think the UI tests are related here. May I ask someone to re-run the tests or help me understand the failure mode?

I'm not sure what's going with the this failure. I suspect it may be because this is an external PR the Percy token may not be exposed. But we can safely ignore this error as that step only uploads UI visual diff testing results, which this PR wouldn't affect.

The changes look good to me, but I will just chat with the rest of the team to make sure I'm not missing anything.

Thanks again for the detailed work!

**Comment by stswidwinski** [comment]
> Yeah, this part of the code is very tricky to test reliably since there's a lot going on, causing tests to be flaky sometimes. [TestClient_WatchAllocs](https://github.com/hashicorp/nomad/blob/14a38bee7bc4386e74157f6a99f3db7382d7e6a5/client/client_test.go#L539-L625) may be a good place to start. I wonder if you setup a test like that and manually mess with the server raft index we could reliably reproduce and test the problem.

Please see the recent commit comments for some more context. What do you think?

**Comment by stswidwinski** [comment]
@lgfa29 please let me know how you'd like to proceed

**Comment by lgfa29** [comment]
Hi @stswidwinski 👋 

Apologies for the delay, but I don't have a good answer yet 😅 

I'm checking with the rest of the team for any additional ideas and I will let you know if something comes up. But since this is such a tricky behaviour to trigger in a test we may need do without out for now.

**Comment by stswidwinski** [comment]
Hey @lgfa29!

Thank you for the update and sorry to keep bugging you. Let me know if I can help in some way or if any part of the bug report (or PR) makes this more confusing. Otherwise, I'll sit tight :)

**Review by schmichael — APPROVED** [review]
I think this is a fine shortcircuit to add, but I am struggling to see where it could fix misbehavior.

> I'm beginning to think that these tests are not quite possible without broader changes to timeout settings.

I think you're correct here. One approach we could take is adding a `rpcQueryWaitTime time.Duration` to the `Client struct` and overriding it during tests to at least exercise the "blocking query timeout" case. At (normal) runtime the zero value should get converted to the default in [`Client.rpc(...)`](https://github.com/hashicorp/nomad/blob/v1.6.1/client/rpc.go#L91-L94), but it's probably best to explicitly set it to [`structs.DefaultBlockingRPCQueryTime`](https://github.com/hashicorp/nomad/blob/v1.6.1/nomad/structs/structs.go#L203) in `client.NewClient(...)`.

The main downside I can see is that this is yet-another-field on the already enormous Client struct, and a test-only one at that. I can't think of a less invasive change we could make to exercise some of this logic though.

I think moving `Client.RPC(...)` and dependencies into a new struct that implements `RPCer interface{ RPC(...) error }` or similar might provide us with a reasonable hook to manipulate RPC behavior in tests, but (a) that's a lot of work, and (b) we could easily mock so much that we're not actually testing realistic runtime scenarios anymore.

Another option would be to refactor the function being changed here into multiple functions and unit test them. This greatly increases the `(b) not testing realistic scenarios` issue above, but there may be opportunities to assert some useful things.

Long story short: I'd love someone to explore better testing options for Client RPCs, but I don't think we need to hold up this PR to wait for them.

**Comment by stswidwinski** [comment]
> I think this is a fine shortcircuit to add, but I am struggling to see where it could fix misbehavior.

I think the basic tenant of the breakage right now is that right now we accept the answer of the Scheduler independent of whether it is new or if it is stale. In the case of staleness we would like to avoid applying the changes that such an answer would imply (killing allocations which are "unexpected" and garbage collecting them). 

The stale replies may exist if the server we are using to get information from (we allow staleness so this is not the leader) times out awaiting for the raft state to catch up with an index which would allow it to process the query correctly. This happens more frequently than one would expect because of the retry logic within the client. The issue submitted is quite verbose about this, but I'm happy to provide any more detail if it's helpful.

The change in a somewhat trivial way discards the data returned by the scheduler are requires that the response corresponds to the minimal index requested.

**Comment by schmichael** [comment]
> I think the basic tenant of the breakage right now is that right now we accept the answer of the Scheduler independent of whether it is new or if it is stale.

Ah you're right! I was thinking this check was sufficient: https://github.com/hashicorp/nomad/pull/18269/files#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09R2396

However that check is only performed for allocs that are observable by the server returning the response. If the server is too stale to see the alloc exists, the Client won't even make it to that check and think the alloc has been GC'd by the server and therefore should be GC'd locally.

Yikes. I'll add a changelog entry and get this merged. Thanks @stswidwinski!
